common/raw_hashtable_metadata_group.h

```diff
@@ -0,0 +1,930 @@
+// auto Operation(...) -> ... {
+//   ... portable_result;
+//   if constexpr (!UseSIMD || DebugChecks) {
+//     portable_result = PortableCode(...);
+//     if (!UseSIMD) {
+//       return portable_result;
+//     }
+//   }
+//   ... result;
+// #if CARBON_NEON_SIMD_SUPPORT
+//   result = NeonCode(...);
+// #elif CARBON_X86_SIMD_SUPPORT
+//   result = X86Code(...);
+// #else
+//   static_assert(!UseSIMD, "Unimplemented SIMD operation");
+// #endif
+//   CARBON_DCHECK(result == portable_result) << ...;
+//   return result;
+// }
+// ```
+class MetadataGroup : public Printable<MetadataGroup> {
+ public:
+  static constexpr ssize_t Size =
+#if CARBON_X86_SIMD_SUPPORT
+      16;
+#else
+      8;
+#endif
+  static_assert(Size >= 8);
+  static_assert(Size % 8 == 0);
+  static_assert(Size <= MaxGroupSize);
+  static_assert(MaxGroupSize % Size == 0);
+  static_assert(llvm::isPowerOf2_64(Size),
+                "The group size must be a constant power of two so dividing by "
+                "it is a simple shift.");
+  static constexpr ssize_t Mask = Size - 1;
+
+  // Each control byte can have special values. All special values have the
+  // most significant bit set to distinguish them from the seven hash bits
+  // stored when the control byte represents a full bucket.
+  //
+  // Otherwise, their values are chose primarily to provide efficient SIMD
+  // implementations of the common operations on an entire control group.
+  static constexpr uint8_t Empty = 0;
+  static constexpr uint8_t Deleted = 1;
+
+  static constexpr uint8_t PresentMask = 0b1000'0000;
+
+  static constexpr bool FastByteClear = Size == 8;
+
+  using MatchIndex =
+#if CARBON_X86_SIMD_SUPPORT
+      BitIndex<uint32_t, /*ByteEncoding=*/false, /*ZeroMask=*/0xFFFF0000>;
+#else
+      BitIndex<uint64_t, /*ByteEncoding=*/true>;
+#endif
+
+  using MatchRange = BitIndexRange<MatchIndex>;
+
+  union {
+    uint8_t bytes[Size];
+    uint64_t byte_ints[Size / 8];
+#if CARBON_NEON_SIMD_SUPPORT
+    uint8x8_t byte_vec = {};
+    static_assert(sizeof(byte_vec) == Size);
+#elif CARBON_X86_SIMD_SUPPORT
+    __m128i byte_vec = {};
+    static_assert(sizeof(byte_vec) == Size);
+#endif
+  };
+
+  auto Print(llvm::raw_ostream& out) const -> void;
+
+  friend auto operator==(MetadataGroup lhs, MetadataGroup rhs) -> bool {
+    return CompareEqual(lhs, rhs);
+  }
+
+  // The main API for this class. This API will switch between a portable and
+  // SIMD implementation based on what is most efficient, but in debug builds
+  // will cross check that the implementations do not diverge.
+  static auto Load(const uint8_t* metadata, ssize_t index) -> MetadataGroup;
+  auto Store(uint8_t* metadata, ssize_t index) const -> void;
+
+  auto ClearByte(ssize_t byte_index) -> void;
+
+  auto ClearDeleted() -> void;
+
+  auto Match(uint8_t present_byte) const -> MatchRange;
+  auto MatchPresent() const -> MatchRange;
+
+  auto MatchEmpty() const -> MatchIndex;
+  auto MatchDeleted() const -> MatchIndex;
```

Comment:
Shouldn't these return `MatchRange` too, like `Match()` and `MatchPresent()`? Or do they only return the first match? If the latter case, please add comments to document.